tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+      with self._lock:
+        while not self._should_wakeup():
+          self._please_sir_i_want_some_more.wait()
+        if self._is_closed:
+          try:
+            self._reader.close()
+            tf.logging.debug('Closed')
+          except Exception as e:
+            self._close_exception = e
+            tf.logging.debug('Close failed: %s', e)
+          self._reader = None
+          self._no_soup_for_you.notify_all()
+          return
+        if not self._should_rebuffer():
+          continue
+        # Calculate a good amount of data to read outside the lock.
+        # The less we have buffered, the less re-buffering we'll do.
+        # We want to minimize wait time in the other thread.
+        x = float(self._buffered)
+        c = float(self._read_ahead)
+        s = BufferedRecordReader.READ_AHEAD_AGGRESSION
+        want = int(min(c - x, s/c * x**s + 1))
+      tf.logging.debug('Waking up to read %s bytes', util.add_commas(want))
+      records = []
+      read_exception = None
+      try:
+        while want > 0:
+          record = self._reader.get_next_record()
+          if record is None:
+            break
+          records.append(record)
+          want -= len(record.record)
+        if self._should_stat():
+          self._stat()
+      except Exception as e:
+        read_exception = e
+      with self._lock:
+        if read_exception is not None:
+          self._read_exception = read_exception
+          self._is_closed = True
+        elif not records:
+          self._is_tapped = True
+        else:
+          for record in records:
+            self._records.append(record)
+            self._buffered += len(record.record)
+        self._no_soup_for_you.notify_all()
+
+  def _should_wakeup(self):
+    return (self._is_closed or
+            self._should_rebuffer() or
+            self._should_stat())
+
+  def _should_rebuffer(self):
+    return (not self._is_tapped and
+            (float(self._buffered) <
+             self._read_ahead / BufferedRecordReader.READ_AHEAD_AGGRESSION))
+
+  def _should_stat(self):
+    return (self._reader.get_offset() > self._size or
+            (self._last_stat <=
+             self._clock() - BufferedRecordReader.STAT_INTERVAL_SECONDS))
+
+  def _stat(self):
+    now = self._clock()
+    size = self._reader.get_size()
+    minimum = max(self._size, self._reader.get_offset())
+    if size < minimum:
+      tf.logging.warning('File shrunk while reading: %s', self.path)
+      size = minimum
+    with self._lock:
+      self._size = size
+      self._last_stat = now
+
+  def __str__(self):
+    return u'BufferedRecordReader{%s}' % self.path
+
+
+def _shorten_event_log_path(path):
+  """Makes an event log path more human readable.
+
+  Returns:
+    Path containing only basename and the first parent directory name,
+    if there is one.
+
+  :type path: str
+  :rtype: str
+  """
+  m = _SHORTEN_EVENT_LOG_PATH_PATTERN.search(path)
+  return m.group(0) if m else None
```

Comment:
I believe the correct type is `optional[str]` but IntelliJ doesn't seem to support it. The [PEP 248](https://www.python.org/dev/peps/pep-0484/) authors didn't put a lot of effort into defining how types work in Python 2.x so there's been a lot of confusion. Even in practice, when I looked at code search on how to do it, I was surprised by how little consensus there was.